wfsim/core.py

```diff
@@ -291,89 +291,106 @@ def __call__(self, instruction):
 
         super().__call__()
 
-    def photon_channels(self, points, n_photons):
+    @staticmethod
+    def get_n_photons(n_photons,positions, s1_light_yield_map, config):
+        if config['detector']=='xenonnt_detector':
+            ly = np.squeeze(s1_light_yield_map(positions),
+                            axis=-1)
+        elif config['detector']=='XENON1T':
+            ly = s1_light_yield_map(positions)
+            ly *= config['s1_detection_efficiency']
+        n_photons = np.random.binomial(n=n_photons, p=ly)
+        return n_photons
 
-        # print(self.config)
-        channels = np.arange(self.config['n_tpc_pmts'])#+1 for the channel map
-        p_per_channel = self.resource.s1_pattern_map(points)
-        p_per_channel[:, np.in1d(channels, self.turned_off_pmts)] = 0
+    @staticmethod
+    def photon_channels(points, n_photons,config,s1_pattern_map):
+        channels = np.arange(config['n_tpc_pmts'])#+1 for the channel map
+        p_per_channel = s1_pattern_map(points)
+        p_per_channel[:, np.in1d(channels, config['turned_off_pmts'])] = 0
         
-        self._photon_channels = np.array([]).astype(int)
+        _photon_channels = np.array([]).astype(int)
         for ppc, n in zip(p_per_channel, n_photons):
-            self._photon_channels = np.append(self._photon_channels,
+            _photon_channels = np.append(_photon_channels,
                     np.random.choice(
                         channels,
                         size=n,
                         p=ppc / np.sum(ppc),
                         replace=True))
+        return _photon_channels
 
-    def photon_timings(self, t, n_photons, recoil_type):
+    @staticmethod
+    def photon_timings(t, n_photons, recoil_type, config):
         if n_photons == 0:
             return
 
-        if (self.config.get('s1_model_type') == 'simple' and 
-           recoil_type in NestId.ALL):
+        if (config.get('s1_model_type') == 'simple' and 
+           recoil_type in NestId._ALL):
             # Simple S1 model enabled: use it for ER and NR.
-            self._photon_timings = np.append(self._photon_timings,
-                t + np.random.exponential(self.config['s1_decay_time'], n_photons))
-            self._photon_timings += np.random.normal(0,self.config['s1_decay_spread'],len(self._photon_timings))
-            return
+            _photon_timings = t + np.random.exponential(config['s1_decay_time'], n_photons)
+            _photon_timings += np.random.normal(0,config['s1_decay_spread'],len(_photon_timings))
+            return _photon_timings
 
-        #TODO Fix the other recoil types to refer to the numbers
+        #TODO Looks nice but don't know if this actually works
         try:
-            self._photon_timings = np.append(self._photon_timings,
-                t + getattr(self, recoil_type.lower())(n_photons))
+            recoil_types = [k for k in vars(NestId) if not k.startswith('_')]
```

Comment:
Here you want to look over `ER`, `NR` and `ALPHA` and **not** over `_ALL` right?

Reply:
yes